### PR TITLE
Add jo

### PR DESCRIPTION
**Pull request by markus1189**
###### Things done:
- [x] Tested using sandboxing (`nix-build --option build-use-chroot true` or [nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS)
- [x] Built on platform(s): NixOS
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"` (No dependencies)
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).
###### More

As this is my first package created it would be great if someone had a closer look ;)
I was especially unsure about `platforms = stdenv.lib.platforms.linux ++ stdenv.lib.platforms.darwin;`

### CODE REVIEW

**Comment by hrdinka** [comment]
Please rename your commits to fit our contributing guidelines:
http://hydra.nixos.org/build/32966504/download/1/nixpkgs/manual.html#chap-submitting-changes

Correct names would be:
`lib.maintainers: add markus1189`
`jo: init at 0.9`

File: pkgs/development/tools/jo/default.nix [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  };
+
+  enableParallelBuilding = true;
+
+  meta = {
+    inherit (s) version;
+    description = ''A small utility to create JSON objects'';
+    license = stdenv.lib.licenses.gpl2Plus;
+    maintainers = [stdenv.lib.maintainers.markus1189];
+    platforms = stdenv.lib.platforms.linux ++ stdenv.lib.platforms.darwin;
```
Comment (hrdinka) [inline pkgs/development/tools/jo/default.nix]: It's okay how you solved this. Have you been concerned about this line cause you are unsure about the chosen platforms or nix syntax?

File: pkgs/development/tools/jo/default.nix [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  inherit buildInputs;
+  src = fetchurl {
+    inherit (s) url sha256;
+  };
+
+  enableParallelBuilding = true;
+
+  meta = {
+    inherit (s) version;
+    description = ''A small utility to create JSON objects'';
```
Comment (hrdinka) [inline pkgs/development/tools/jo/default.nix]: Just a unimportant nitpick: This isn't a multi-line string nor are there any escaping issues so you can use normal double quotes (") here as well.

**Comment by joachifm** [comment]
This style of expression is a bit unusual. Is the source information actually autogenerated, and if so how?

File: pkgs/development/tools/jo/default.nix (hunk 1 above)
Comment (joachifm) [inline pkgs/development/tools/jo/default.nix]: I'm pretty sure you can do `platforms.unix` or even `platforms.all` here, seeing as they provide packages for OpenBSD as well. Can't see anything but standard C/POSIX includes.

File: pkgs/development/tools/jo/default.nix [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+let
+  s = # Generated upstream information
+  rec {
+    baseName="jo";
+    version="0.9";
+    name="${baseName}-${version}";
+    url="https://github.com/jpmens/jo/releases/download/v${version}/${baseName}-${version}.tar.gz";
+    sha256="07r4m9g8696gvgz9z9m2zk83nc1ifa5siiyqplp6ijg5d5sd2mz7";
+  };
+  buildInputs = [];
```
Comment (joachifm) [inline pkgs/development/tools/jo/default.nix]: This is redundant

File: pkgs/development/tools/jo/default.nix [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+in
+stdenv.mkDerivation {
+  inherit (s) name version;
+  inherit buildInputs;
+  src = fetchurl {
```
Comment (joachifm) [inline pkgs/development/tools/jo/default.nix]: The autogenerated tarballs served by github sometimes change (compression and whatnot), resulting in annoying checksum changes. `fetchFromGitHub` (or even `fetchzip`) should be more stable than `fetchurl`.

**Comment by hrdinka** [comment]
@joachifm yes this is auto generated. Have a look over here:

https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/upstream-updater/update-walker.txt

There are a few more of these in nixpkgs. Search for *.upstream files. I would like it more if it would produce a more common output...

File: pkgs/development/tools/jo/default.nix [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    inherit (s) url sha256;
+  };
+
+  enableParallelBuilding = true;
+
+  meta = {
+    inherit (s) version;
```
Comment (joachifm) [inline pkgs/development/tools/jo/default.nix]: I believe `meta.version` is deprecated

**Comment by joachifm** [comment]
@hrdinka I'm not convinced that update-walker.sh could generate that output simply from the project's github front page, but I'm not really motivated to try it either, so maybe it did. Note, however, that there's only one contributor who was ever touched update-walker, I don't think it's been generally accepted.

File: pkgs/development/tools/jo/default.nix (hunk 5 above)
Comment (hrdinka) [inline pkgs/development/tools/jo/default.nix]: I think this is needed by the update walker script. There even is a new `updateWalker` attribute as described in the [manual](https://nixos.org/nixpkgs/manual/#sec-standard-meta-attributes) (why do I miss such stuff...).

**Comment by hrdinka** [comment]
@joachifm I don't like this either. The output looks totally quirky and it introduces just another build description file with a different less flexible syntax (`.upstream` instead of nix—great). It further makes judging about expressions hard because I assume that all manual changes (like the redundant `buildInputs`) will be undone by the next auto update. I don't think that this is generally accepted either. I have first heard about it in this pr https://github.com/NixOS/nixpkgs/pull/13819.

@markus1189 I am curious why you have decided to use `update-walker`. How have you heard of it?

File: pkgs/development/tools/jo/default.nix (hunk 5 above)
Comment (joachifm) [inline pkgs/development/tools/jo/default.nix]: See https://github.com/NixOS/nixpkgs/pull/12156

**Comment by hrdinka** [comment]
@joachifm Let's wait for @markus1189 response. Maybe there is a good reason or discussion about why/when to use this tool. If there is non my vote is for rewriting this as a "proper" nix derivation.

**Comment by markus1189** [comment]
I copied the `default.nix` from `jq` ;)  Thanks for all the comments, I'll have a look now.

**Comment by markus1189** [comment]
Okay I think I addressed all issues ;)

File: pkgs/development/tools/jo/default.nix [hunk 6]
```diff
@@ -0,0 +1,27 @@
+{stdenv, fetchFromGitHub, autoreconfHook}:
+
+stdenv.mkDerivation rec {
+  name="${pname}-${version}";
```
Comment (joachifm) [inline pkgs/development/tools/jo/default.nix]: For consistency, add spaces around `=`.

File: pkgs/development/tools/jo/default.nix [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  pname="jo";
```
Comment (joachifm) [inline pkgs/development/tools/jo/default.nix]: If this is to be used only once, you may inline it into `name`.

**Comment by joachifm** [comment]
@markus1189 looks good, only nitpicks remain, as far as I can tell

File: pkgs/development/tools/jo/default.nix [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    owner = "jpmens";
+    repo = "jo";
+
+    rev = "v${version}";
+    sha256="0hv1spbbaca0bvddn0wxk06mm95xnx4zfi99q79hxj1grxvpj1np";
+  };
+
+  enableParallelBuilding = true;
+
+  buildInputs = [autoreconfHook];
```
Comment (joachifm) [inline pkgs/development/tools/jo/default.nix]: For correctness, use `nativeBuildInputs` for build tools.

**Comment by markus1189** [comment]
:+1:

File: pkgs/development/tools/jo/default.nix [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    owner = "jpmens";
+    repo = "jo";
+
+    rev = "v${version}";
+    sha256="0hv1spbbaca0bvddn0wxk06mm95xnx4zfi99q79hxj1grxvpj1np";
+  };
+
+  enableParallelBuilding = true;
+
+  nativeBuildInputs = [autoreconfHook];
```
Comment (hrdinka) [inline pkgs/development/tools/jo/default.nix]: If you are already on it you might want add spaces around list entries as well `[ autoreconfHook ]` to better fit our coding style.

File: pkgs/development/tools/jo/default.nix [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+  meta = with stdenv.lib; {
+    description = "A small utility to create JSON objects";
+    homepage = "https://github.com/jpmens/jo";
```
Comment (hrdinka) [inline pkgs/development/tools/jo/default.nix]: You can omit the quotes here. URIs are actually native types in nix.

**Comment by hrdinka** [comment]
@markus1189  :+1: This looks much better now.

@joachifm do you know the current state of update walker? I am thinking of opening another pr to remove this tool. Many expressions shipping an `.upstream` file do not seem to use it anymore and the ones who do look like this commit before the rewrite. This might be one of the things which are around since ever and no one dared to remove (or update) it yet.

**Comment by joachifm** [comment]
@hrdinka sounds like a good idea. I also think that it's still in there only because nobody has moved to delete it. In light of the recent discussion about the proliferation of autoupdaters, such a PR might very well be accepted.

**Comment by hrdinka** [comment]
@joachifm Thanks for your encouragement. As a new NixOS member I wasn't not sure if people would accept such changes or if I could even piss off some of our top contributors. There isn't much guidance in form of community guidelines or etiquette and I am not always sure about what's okay to do and what's not. I guess many contributors have these problems and recent discussions about commit reverts or the "NixOS chaos" manifest this.

**Comment by joachifm** [comment]
@hrdinka Proposing a change and arguing its merits is never wrong, in my opinion, and I'm pretty sure most people here would agree.

**Comment by markus1189** [comment]
Okay changed to `1.0` which was just released and also addressed the remaining nitpicks ;)

**Comment by hrdinka** [comment]
Built & tested locally (NixOS x86_64). Thanks for your contribution :+1:

**Comment by 7c6f434c** [comment]
@joachifm @hrdinka Actually, `upstream-walker` can update usual-style `fetchurl` expressions fine (and I also use it for that), it just doesn't generate them.

The version that put updater metadata in Nix caused _more_ complaints…

And note that `update-walker` only pretends to be declarative — it is declarative for simple things, but imperative for more complex things (see `atop` package, for example, or `sleuthkit`); so putting it into Nix attributes would reduce the expressiveness.

**Comment by joachifm** [comment]
@7c6f434c did you ever push to adopt update-walker across nixpkgs? I'm interested to read past discussions, but google is failing me, all I can find are people asking what default.upstream files are.

**Comment by hrdinka** [comment]
@7c6f434c Thanks for your explanation. It took me some while to find this out by my own. I am also interested in update-walker's acceptance within nixpkgs. I am currently using my own set of external scripts for such stuff and I am curious if I should better switch to update-walker.

**Comment by 7c6f434c** [comment]
I have never tried to convince people to adopt the update-walker across
NixPkgs; it doesn't look like there is a consensus that a working 
auto-updating infrastructure is needed (look at monitor.nixos.org fate,
for example), and so I use it for packages I actually try to keep up to
date. Of course, I try to answer questions about it, too; I was even
shamed into writing a brief overview you have seen.

I do not believe in global adoption; it looks like most people are OK 
ignoring .upstream files, unlike things that split the actual Nix 
expressions into two files just for update reasons; I am willing to 
implement or help implementing any specific enough suggestions that 
would help you to also use it, of course.

**Pull request merged**